# <center>MC906/MO416 - Introduction to Artificial Intelligence</center>
# <center>Institute of Computing - Unicamp</center>
# <center>Prof. Esther Colombini</center>

### <center>Eduardo S. Ito (RA159086)</center>
### <center>Lucas Peres  (RA 265193)</center>
### <center>Thales E. Nazatto (RA 074388)</center>

# Conteúdo

1. Introdução
    1. Descrição do problema
2. Metodologia
    1. Modelagem da solução
        1. Requisitos
        2. Implementação
        3. Problema, agentes e ambiente
    2. Algoritmos
        1. Busca não-informada
        2. Busca informada
        3. Busca local (*Hill Climbing*)
3. Resultados
4. Conclusão
5. Referências

# Introdução
**Pac-man** é um jogo para *Arcades* lançado em 1980 pela **Namco**, criado por Toru Iwatani e ilustrado na Figura 1. Nele, o jogador controla uma bolinha amarela (o *Pac-Man* do título) dentro de um labirinto que tem como objetivo comer todas as bolinhas brancas existentes nele enquanto é perseguido por 4 fantasmas de nomes *Blinky*, *Pinky*, *Inky* e *Clyde*. Na época, Iwatani criou o jogo para atrair pessoas de ambos os sexos, uma vez que os gêneros de jogos lançados nessa época, como jogos de guerra e esportes, eram majoritariamente pensados para agradar o público masculino. Tal estratégia se provou um grande sucesso, e *Pac-Man* se tornou uma das mais lucrativas franquias de todos os tempos, gerando mais de US$ 14 bilhões em receita até o ano de 2016 **[1]**.

<img src="https://thumbs.web.sapo.io/?W=775&H=0&delay_optim=1&webp=1&epic=NGRiQ8y+7rDb2cocSbMwl47ynqxTFYLIDm2IhR5IZK5BFQooKY19ghBWe//2n+M5C2MgCiqhexDuDLBr/f/MrofnNA==" height="240" width="320" alt="Tela do jogo Pac-Man" title="Tela do jogo Pac-Man" />
<center><strong>Figura 1.</strong> Tela do jogo <i>Pac-Man</i></center>

Uma dos grandes inovações para a época foi sua Inteligência Artificial. Cada fantasma se comportava de maneira diferente de acordo com os movimentos realizados pelo jogador, que poderiam ter a possibilidade de encurralá-lo. Enquanto *Blinky* perseguia *Pac-Man* diretamente, *Pinky* e *Inky* tentavam se posicionar a sua frente e *Clyde* alternava seus estados entre perseguir o *Pac-Man* e fugir dele. Tais comportamentos fizeram com que *Pac-Man* fosse um caso de estudos no ramo da Inteligência Artificial.



Neste relatório, será mostrada uma modelagem do problema determinado em **[2]** utilizando 5 algoritmos de busca no contexto deste jogo, sendo 2 de busca não-informada (*Breadth-First Search* e *Depth-First Search*), 2 de busca informada (*Greedy Best-First Search* e *A*\*) e 1 de busca local (*Hill Climbing*). Primeiro será detalhada como foi feita a modelagem do problema e as soluções aplicadas. Depois, serão feitos experimentos e dicussões para avaliar o comportamento dos algoritmos e, no final, as conclusões destes experimentos.

## Descrição do problema

Conforme determinado em **[2]**, o problema consiste em implementar 5 algoritmos diferentes de busca, sendo 2 de busca não-informada, 2 de busca informada (com 2 heurísticas diferentes) e 1 de busca local usando o *Pac-man* como tema, tendo como objetivo comer todas as bolinhas presentes no labirinto, caso consiga evitar os fantasmas. Dentre esses, é necessário avaliar critérios como otimalidade, completude e custo computacional para definir qual a melhor solução dentre elas.

O labirinto é representado por um grid, também ilustrado na Figura 2, cujo tamanho e forma devem ser definidos. Deve ser definido também, além das posições das bolinhas que o *Pac-man* deve comer durante a execução do algoritmo, as posições de 3 fantasmas (que serão estáticos) e uma posição final de parada. É necessário especificar também a representação dos estados, as ações realizadas, o teste realizado para se chegar ao objetivo, o custo **g(x)** do caminho percorrido e as heurísticas utilizadas.

<img src="https://shaunlebron.github.io/pacman-mazegen/img/origmaps_2x_print.png" height="360" width="480" alt="Tela do jogo Pac-Man" title="Tela do jogo Pac-Man" />
<center><strong>Figura 2.</strong> Exemplos de labirintos semelhantes <strong>[6]</strong></center>

# Metodologia

In [14]:
# CHAPTER 1: REQUIREMENT SPECIFICATION

# 1. Goal

# [R1] Two search methods without information
# [R2] Two informed search methods with 2 distinct heuristics
# [R3] One local search method

# The work consists of finding an adequate solution to the chosen problem, evaluating it according to: computational
#cost, completeness, optimality. Your are required to clearly define:

# [R4] How the problem was modeled
# [R5] Implementation specifics and restrictions

# The work consists of ending an adequate solution to the chosen problem, evaluating it according to: [R6]computational
# cost, [R7] completeness, [R24] optimality. Your are required to clearly define:
# [R8] How the problem was modeled
# [R5] Implementation specifics and restrictions

# 2. Problem
# Pac-man is one of the most popular Arcade games, still played nowadays. In the game, there is a maze where Pac-man
# has to act to collect as many pieces as possible without being caught by one of the ghosts. The maze is defined in
# terms of a grid, as presented in Figure 1. The black areas are non-traversable walls, the grey path represents elements
# that should be caught by the agent, and the white areas are traversable areas. Your group should define additional
# elements, such as:

# [R10] The size and shape of the maze
# [R11] A final goal position
# [R12] The static positions of 3 Ghosts in the scene (the positions do not change during the game)
# [R13] Pac-man's initial position

# To solve the problem, you have to specify:

# [R14] The state representation
# [R15] The set of actions
# [R16] The objective state test
# [R17] The cost of the path (g(x))
# [R18] The heuristics used

# The system must be evaluated according to the quality of the solutions found and a critical evaluation is expected on
# the relationship between adopted parameters x solution performance. Graphs and tables representing the evolution of
# the solutions are expected. Additional comparisons with the literature are welcome, although they are not mandatory.
# To evaluate the results you might change the following elements:

# [R19] Pac-man's initial position
# [R20] Goal position
# [R21] State discretization (maze size and configuration)
# [R22] Ghosts' position

# [R23] It is important to remember that we will not perform an online search. As the problem is completely observable and
# deterministic, Pac-man is going to reason about the solution before actually adopting it.

# CHAPTER 2: MODELING PROPOSAL

# Chapter 2.1 [R4] How the Problem was modelled?
# The Berkeley algorithm algorithm was used as design base for our implementation.
#        Artificial Intelligence in Pacman Class material offered in the course 
#        "Artificial Intelligence (2019 Spring) of NTU" at url https://github.com/andi611/Pacman-With-AI-Python.
#        The reason was to comply with [R10]..[R13], it was understood that graphical interface is required, at initial
#        stage of project. Despite, the classroom on April 29th, it was clarified that it was not required.
#
# The reason why requirement [4] Python implementation of algorithms from Russell And Norvig's "Artificial Intelligence
#        - A Modern Approach"  https://github.com/aimacode/aima-python, was not used, is that Berkeley algorithm and Node 
#        library as used by Russell and Norvig are incompatible. The state used by them differs on coordinate axis
#        and the integration of graphical interface and list of actions produced by search algorithm could take very long,
#        and could become unfeasible to comply with mostly of requirements, in time. Berkeley algorithms use Game library,
#        and Russell and Norvig algorithms use Node library. More differences will be explained on this report.This is also
#        for sake of requirement [5] reporting limitations.
#
# Our  Project Repository is stored at folder Prototype3, only this, at github address https://github.com/edbkei/MO416AI
#
# For sake of operationally, the command line inherited "partly" by Berkeley algorithm our pacman use cases, do the following: 
#        !python pacman.py -l layoutMO416b -p Search2Agent -a fn=<fn> -z <z>
#        -l layout parameter. We use only the layoutMO416b, nevertheless others are allowed, still.
#        -p agent parameter. We use only Search2Agent as the manager for accessing search methods existing in search2.py,
#                            accordingly to problem concept and parameter -a selection. SearchAgents.py and search.py
#                            were used as design base.
#        -a search method. We use dfs (depth-first search ), ucs (uniform cost search), bfs (breadth-first search), 
#                          astar (A* with Manhattan Heuristic), gbfs (greedy best fisrt search with Manhattan Heuristic), 
#                          astare (A* with Euclidean Heuristic), gbfes (greedy best fisrt search with Euclidean Heuristic), 
#                          and hcs (hill climbing search, local search),
#
#                          The design base had already dfs, ucs, bfs, astar (with null heuristic), not existed before
#                          gbfs neither hcs.
#
#        -z zoom. Numberic value (0.1..0.9)
#
# Conversion from Python 2.7 to Python 3.4. Fix several incompatibility as print format, and library tkint.
#        https://docs.python.org/3/library/tkinter.html. This is also to report constraint as requested by requirement [5].
#
# We customized Berkeley algorithm, accordingly to following:
#       The implementation of requirement [R10] with maze were done by customization of maze generated by
#       https://shaunlebron.github.io/packman-mazegen. We included initial pacman position, only an unique food,
#       3 ghosts located in random positions in the maze. Our maze is called layoutMO416b.lay, as the format understood by
#       Berkeley format. It is located at ../MO416AI/Prototype3/layout.
#       The requirement [R10] is implemented with maze represented by the file. 
#       The requirement [R10] about size and shape are inherited by Berkeley algorithm. The command parameter -z
#       makes zoom possible. 
#
#       The requirement [R11] is implemented as output in search2Agents.py
#       The requirement [R12] is implemented by means of pause in the dynamicity of ghost on script pacman.py
#       The requirement [R13] is implemented as output in search2Agents.py
#       The requirement [R16] is implemented as output in search2Agents.py
#       The requirement [R17] is implemented as output in search2Agents.py
#       The requirement [R18] is complied with euclidean and Manhattan heuristic from design base, but not used
#       in our use cases. We implemented manhattan and euclidean heuristic (h) for A* and gbfs, in hard code way.
#       The original algorithm of A* considers the stop condition f(x)=g(x)+h(x) with null heuristic, 
#       and we did the same algorithm as A* but with f(x) = h(x) for gbfs. We implemented in search2Agents.py and search2.py.
#       The hill climbing search was created from the scratch.
#       The requirement [R20] is implemented in the class PositionSearchProblem2, in search2Agents.py, using the
#       the methods gameState.getFood and gamestate.hasFood. Now, it is possible to put food, but only one, in any place of
#       the maze.
#
#       New function codes to make possible to specify parameter value gbfs and hcs in the command line were introduced. 
#       Implemented in search2.py.
#
#      The requirement [R14], also part of requirement [8], is implemented by the concept of problem used by Berkeley
#      algorithm is PositionSearchProblem2 (similar to original PositionSearchProblem, but with additions herein explanained).
#      It is a tuple of (initial state position of the pacman, the goal position of the food, 
#      action (if next step of pacman is east, west, north, south), cost of 1 for each step in the maze)
#
#      When a command line is issued, search2Agents is invoked, and according to function code, the PositionSearchProblem2 
#      class is invoked to build an actions list produced by search method (in search2.py). Owning this list, 
#      the class search2Agents can send it to game graphics interface.
#      
#      For sake of clarity, despite inherited algorithm from Berkeley, the used classes will be represented in class diagram,
#      by customization of pyreverse output.
#
#      And the used search methods are described in high level algorithms. See more in modelingProposal at item [11].
#      
#      The requirement [21] is implemented in design base of Berkeley algorithm, and -z parameter (zoom). It is also
#      implemented layout.py and layout data in format .lay.
#
#      The requirement [R23] is already implemented from design base of Berkeley algorithm. Basically, depending on use case 
#      defined by commmand line, where search method, and layout chosen. The search methods return actions list where pacman
#      should pass by from start state to goal state. This actions list is submitted to graphical interface that runs pacman
#      game.
#
#      Modeling version information is given here https://github.com/edbkei/MO416AI, at README.


# Methods used in search2.py. It complements the requirement [R14] to represent the state representation
# and requirement [R15] The set of actions. Also it also complies to requirement [8].
#
import os
cwd = os.getcwd()
print(cwd)
cwd=cwd+'/modelling/modellingProposal'
f = open(cwd, 'r')
file_contents = f.read()
print (file_contents)

f.close()

/home/lucas/mo416/MO416AI/Project1


1. Modeling Proposal.

function BREADTH_FIRST_SEARCH(problem) returns a list of actions towards the pacman goal {
# The aim is to return a list of actions pacman to reach the goal (a food) by
# searching the shallowest nodes or states in the search tree first.
#
input: problem (a tuple of start_state, action, cost, goalState)
# node or state has the same meaning for pacman game
local variables: current_state (a node) and next_state (a neighbor node)

queue := util.Queue();
trace := empty list;
seenList := empty list;

start_state := CREATE-NODE(START-STATE[problem]);
initial push on queue with start_state;
Append (start_state) into seenList;

while queue is not empty do

   current_state := pop from queue;
   
   if goal state from problem is reached at current_state then  
      break;
   end if;
   
   successors := getSucccessors(current_state) from problem;
   
   for successor in successors do
       next_state := state of successor;
	   next_a

## Modelagem da solução

### Requisitos

Dada a descrição do problema, foram levantados os seguintes requisitos para a conclusão deste projeto:

- **[R1]** 2 algoritmos de busca não-informada
- **[R2]** 2 algoritmos de busca informada (com 2 heurísticas diferentes)
- **[R3]** 1 algoritmo de busca local
- **[R4]** Definir como o problema é modelado
- **[R5]** Especificações e restrições da aplicação
- **[R6]** Avaliar custo computacional dos algoritmos
- **[R7]** Avaliar completude dos algoritmos 
- **[R8]** Avaliar otimalidade dos algoritmos 
- **[R9]** Definir o tamanho e formato do labirinto
- **[R10]** Definir uma posição final de parada
- **[R11]** Definir as posições de 3 fantasmas (que serão estáticos)
- **[R12]** Definir a posição inicial do *Pac-man*
- **[R13]** Especificar representação dos estados
- **[R14]** Especificar as ações realizadas
- **[R15]** Especificar o teste realizado para se chegar ao objetivo
- **[R16]** Especificar o custo **g(x)** do caminho percorrido
- **[R17]** Especificar as heurísticas utilizadas
- **[R18]** A posição inicial do *Pac-man* pode variar de acordo com o problema
- **[R19]** A posição final de parada pode variar de acordo com o problema
- **[R20]** Discretização dos estados (tamanho e configuração do labirinto)
- **[R21]** As posições dos fantasmas podem variar de acordo com o problema
- **[R22]** A busca não será *online*. O problema é completamente observável e determinístico

### Implementação

Para a implementação, foi escolhido em vez da biblioteca AIMA **[3]** o código da Universidade de Berkeley **[4]**. Inicialmente a principal motivação de fazer a troca reside no fato que o código já possui as funções gráficas para exibição do problema, fazendo com que o foco das mudanças seja apenas na modelagem e nos algoritmos. Outras vantagens notadas é que a adaptação deste código para a representação de estados e ações foi mais simples e ele já possuía ferramentas para testes do objetivo. Entretanto, após informações futuras foi comprovado que não seria estritamente necessário e, como o Python 3 foi escolhido para a implementação, foram necessárias adaptações devido a diferenças na biblioteca *tkint* **[5]**. 

Originalmente, o algoritmo de Berkeley já tem um agente pacman, um agente de busca (searchAgent), um agente de teclado (keyboardAgent) e agente(s) fantasma(s) (ghostAgent), que pode ter várias réplicas, comida(s) (food) e um layout de um labirinto (maze). A ideia básica é que um layout pode ser construída de forma flexível com formatos genéricos, sobre os quais se colocam os muros (representados por %), a comida (representado por . em quaisquer posição), o pacman (representado por P, em quaisquer posições), os fantasmas (representados por um ou mais G, em quaisquer posições). O objetivo do pacman é comer todas as comidas do labirinto, mesmo que seja apenas uma unitária, e assim ganhar o jogo, o(s) fantasma(s) vai/vão tentar atacar o pacman de forma dinâmica e assim o conseguir, o pacman perde o jogo. O problema do algoritmo foi modelado de forma que há um estado de início do pacman (posição no maze), um objetivo (ou estado final, ou objetivo, ou goal) - originalmente hardcoded, funcão de sucessão e função de custo. Por comando de linha se determina qual o agente de busca (searchAgent) e o algoritmo de busca (não informado, informado, local) será utilizado, mas não implementa nenhuma heurística e parâmetro para apresentação gráfico no labirinto (i.e.o zoom). O algoritmo tem duas partes principais, a determinação da árvore de estado, que é o caminho que será percorrido pelo pacman ao longo do labirinto, em forma de lista de ações (EAST, WEST, NORTH, SOUTH). O qual será utilizado como entrada na segunda etapa para a operação do pacman na representação gráfica. O pacman pode ter uma lista de ações até o objetivo final, mas o planejamento pode ser interrompido por um agente fantasma a qualquer momento e assim ser comido e perder o jogo. 

Não houve uma mudança significativa na segunda etapa onde a lista de ações é submetida à ação da parte gráfica. Apenas foi introduzido um código para pausar os agentes, ora dinâmico, dos fantasmas. Ainda assim, com capacidade de matar o agente pacman. Foram extendidos a atual capacidade do agente de busca para cumprir com os requisitos do Projeto 1.

A principal mudança realizada em relação ao código original os arquivos **searchAgents.py** and **search.py**, que possuíam os códigos para modelagem dos agentes e para os algoritmos de busca, foram substituídos por versões customizadas, batizados de **search2Agents.py** and **search2.py** respectivamente. As configurações e usabilidade também foram herdadas do código da Universidade de Berkeley, sendo executadas na seção de Experimentos e discussão e utilizando estas *flags*:

- **-l** - Determina o *layout* de labirinto a ser utilizado
- **-p** - Determina o agente inteligente que controlará o *Pac-Man*
- **-a** - Determina os parâmetros dos algoritmos utilizados. É possível selecionar o algoritmo utilizado (**fn**) e a heurística (**heuristic**)
- **-z** - Zoom. Determina a escala da janela a ser aberta para visualização do jogo (de 0.1 a 0.9)

Os labirintos são gerados em arquivos de texto, em arquivos de extensão **.lay**. Nele, o programa reconhece como códigos os seguintes caracteres:

- **%** - Parede
- **.** - Bolinha
- **o** - Bolinha com poderes
- **G** - Fantasma
- **P** - *Pac-man*

Tal padrão permite um desenvolvimento mais flexível e intuitivo dos labirintos, podendo colocar o *Pac-man*, os fantasmas e as bolinhas em diversas posições e com tamanhos e formatos variáveis de labirintos. Devido aos requisitos do projeto, foram colocadas as seguintes restrições:

- Como o objetivo do problema prioriza a avaliação dos algoritmos, o problema foi modelado com o objetivo mais simples: Encontrar a única bolinha que está presente nos mapas propostos.
- Não foram colocadas bolinhas com poderes nos labirintos. Isso dá mais ênfase a busca, pois o *Pac-man* sempre perderá o jogo se o algoritmo escolher o caminho onde está um dos fantasmas.
- São colocados 3 fantasmas em posições determinadas, porém distintas.
- O agente inteligente dos fantasmas é desligado durante o jogo.
- Para simplificar os *layouts* e testes dos algoritmos, o labirinto possui o formato retangular e fechado.
- O labirinto é feito em **[6]**, colocado em arquivo e possui um tamanho de 28 linhas x 31 colunas
- O *Pac-man* é colocado em uma posição determinada.

### Problema, agentes e ambiente

Como o labirinto é representado por um grid, os estados foram representados por tuplas no formato **(x,y)** e as ações são determinadas pelas direções **['South', 'West', 'East', 'North']**. Uma implementação simples do problema na biblioteca AIMA pode ser verificada no código abaixo:

Nele, a classe **SimplePacmanProblem** é uma classe filha da classe **Problem**, que é passada para os algoritmos de busca encontrarem o melhor caminho. Como parâmetros em seu construtor há um valor inicial, final, obstáculos e heurística. **action_cost** é a função heurística utilizada para algoritmos de busca informada (ou **g(x)**), **h** é a função heurística utilizada para o *A*\* (ou **h(x)**) e **value** é a função heurística utilizada para algoritmos de busca locais (ou **g(x)**). No AIMA, o heurística nos algoritmos de busca locais em seu código original é realizada em direção a um mínimo local. O valor de **directions** seria semelhante às ações que o *Pac-man* executaria no caminho, fazendo estados e ações serem tuplas no formato **(x,y)** e praticamente não tendo diferenciação.

A classe semelhante no código utilizado neste projeto é a **PositionSearchProblem2**, que é uma classe filha da classe **SearchProblem2**, a qual é passada para os algoritmos de busca encontrarem o melhor caminho. Como parâmetros em seu construtor há um valor inicial, final, heurística e o estado do jogo, que guarda as posições das paredes, fantasmas, das bolinhas e do *Pac-man*. **getCostOfActions** é a função utilizada para determinar os estados através da função heurística para os algoritmos de busca informada e local (dependendo dos parâmetros utilizados pode ser **g(x)** ou **h(x)**). **getSuccessors** é a função que disponibiliza os nós sucessores da busca, semelhante a função **actions** presente na AIMA. Há ainda a função **isGoalState**, semelhante a função **goal_test** da classe **Problem**. A classe **Search2Agent** faz o papel de agente inteligente e a ligação entre problema, algoritmo de busca, estado do jogo e heurística.

Um diagrama de classes com todas as classes presentes no código pode ser visto no PDF abaixo:


In [1]:
# This .pdf file shows class diagram that implements how the problem was modeled using game graphical interface. It is part
# of how requirement [R4] and requirement [8] was modeling from Berkeley algorithm design base, and modification done
# to comply with MO416 Project 1 requirements.
import os
cwd = os.getcwd()
cwd=cwd+'/modelling/modellingClasses.pdf'
print(cwd)
import webbrowser
webbrowser.open(r''+os.path.join('modelling','modellingClasses.pdf'))
webbrowser.open(cwd)

C:\Users\Thales E. Nazatto\Documents\Pós\MO416A\MO416AI\Project1/modelling/modellingClasses.pdf


True

## Algoritmos

### Busca não-informada

#### Depth-First Search

In [80]:
# [R1.1] Show dfs (depth first search) method.
!python pacman.py -l easyLayout -p Search2Agent -a fn=dfs -z .6
#!python pacman.py -l normalLayout -p Search2Agent -a fn=dfs -z .6
#!python pacman.py -l hardLayout -p Search2Agent -a fn=dfs -z .6

[Search2Agent] using function dfs
[Search2Agent] using problem type PositionSearchProblem2
[R12] Initial position of pacman is (23, 29)
[R10] Final goal position is (1, 3)
[R11] Ghost Positions is/are [(9, 20), (15, 10), (26, 7)]
Number of foods is 1
[R15] has the game food? False
[R16] Path found with total cost g(x) of 170 in 0.03629422187805176s
[R13] Search nodes expanded: 253
[R13] Nodes visited: [(23, 29), (22, 29), (21, 29), (20, 29), (19, 29), (18, 29), (17, 29), (16, 29), (15, 29), (14, 29), (13, 29), (12, 29), (11, 29), (10, 29), (9, 29), (8, 29), (7, 29), (6, 29), (5, 29), (4, 29), (3, 29), (2, 29), (1, 29), (1, 28), (1, 27), (1, 26), (1, 25), (1, 24), (1, 23), (1, 22), (1, 21), (1, 20), (2, 20), (3, 20), (4, 20), (5, 20), (6, 20), (6, 19), (6, 18), (6, 17), (6, 16), (5, 16), (4, 16), (3, 16), (2, 16), (1, 16), (1, 17), (1, 18), (1, 19), (3, 15), (3, 14), (3, 13), (2, 13), (1, 13), (1, 12), (1, 11), (1, 10), (2, 10), (3, 10), (3, 9), (3, 8), (3, 7), (2, 7), (1, 7), (4, 7), (

#### Breadth-First Search

In [81]:
# [R1.2] Show bfs (breadth first search) method.
!python pacman.py -l easyLayout -p Search2Agent -a fn=bfs -z .6
#!python pacman.py -l normalLayout -p Search2Agent -a fn=bfs -z .6
#!python pacman.py -l hardLayout -p Search2Agent -a fn=bfs -z .6

[Search2Agent] using function bfs
[Search2Agent] using problem type PositionSearchProblem2
[R12] Initial position of pacman is (23, 29)
[R10] Final goal position is (1, 3)
[R11] Ghost Positions is/are [(9, 20), (15, 10), (26, 7)]
Number of foods is 1
[R15] has the game food? False
[R16] Path found with total cost g(x) of 24 in 0.029007434844970703s
[R13] Search nodes expanded: 153
[R13] Nodes visited: [(23, 29), (23, 28), (24, 29), (22, 29), (23, 27), (25, 29), (21, 29), (23, 26), (26, 29), (20, 29), (22, 26), (26, 28), (19, 29), (21, 26), (26, 27), (18, 29), (21, 25), (26, 26), (17, 29), (21, 24), (26, 25), (16, 29), (21, 23), (26, 24), (15, 29), (22, 23), (20, 23), (26, 23), (15, 28), (14, 29), (23, 23), (19, 23), (26, 22), (15, 27), (13, 29), (23, 22), (18, 23), (26, 21), (15, 26), (12, 29), (23, 21), (18, 24), (18, 22), (17, 23), (26, 20), (16, 26), (12, 28), (11, 29), (23, 20), (18, 25), (18, 21), (16, 23), (26, 19), (25, 20), (17, 26), (12, 27), (10, 29), (24, 20), (22, 20), (18,

### Busca informada

#### A*

In [82]:
# [R2.1] Show A* and Manhattan heuristic method.
!python pacman.py -l easyLayout -p Search2Agent -a fn=astar,heuristic=manhattanHeuristic -z .6
#!python pacman.py -l normalLayout -p Search2Agent -a fn=astar,heuristic=manhattanHeuristic -z .6
#!python pacman.py -l hardLayout -p Search2Agent -a fn=astar,heuristic=manhattanHeuristic -z .6

[Search2Agent] using function astar and [R16] heuristic manhattanHeuristic
[Search2Agent] using problem type PositionSearchProblem2
[R12] Initial position of pacman is (23, 29)
[R10] Final goal position is (1, 3)
[R11] Ghost Positions is/are [(9, 20), (15, 10), (26, 7)]
Number of foods is 1
[R15] has the game food? False
[R16] Path found with total cost g(x) of 24 in 0.013302087783813477s
[R13] Search nodes expanded: 49
[R13] Nodes visited: [(23, 29), (23, 28), (22, 29), (23, 27), (21, 29), (23, 26), (20, 29), (22, 26), (19, 29), (21, 26), (18, 29), (21, 25), (17, 29), (21, 24), (16, 29), (15, 29), (15, 28), (14, 29), (15, 27), (13, 29), (15, 26), (12, 29), (12, 28), (11, 29), (12, 27), (10, 29), (12, 26), (9, 29), (11, 26), (8, 29), (10, 26), (7, 29), (9, 26), (6, 29), (9, 25), (9, 24), (24, 29), (21, 23), (16, 26), (5, 29), (9, 23), (20, 23), (8, 23), (19, 23), (7, 23), (18, 23), (6, 23), (18, 24), (17, 23), (6, 24)]
[R13] Solution states: 25 - [(23, 29), (22, 29), (21, 29), (20, 29)

In [83]:
# [R2.1] Show A* and Euclidean heuristic method.
!python pacman.py -l easyLayout -p Search2Agent -a fn=astar,heuristic=euclideanHeuristic -z .6
#!python pacman.py -l normalLayout -p Search2Agent -a fn=astar,heuristic=euclideanHeuristic -z .6
#!python pacman.py -l hardLayout -p Search2Agent -a fn=astar,heuristic=euclideanHeuristic -z .6

[Search2Agent] using function astar and [R16] heuristic euclideanHeuristic
[Search2Agent] using problem type PositionSearchProblem2
[R12] Initial position of pacman is (23, 29)
[R10] Final goal position is (1, 3)
[R11] Ghost Positions is/are [(9, 20), (15, 10), (26, 7)]
Number of foods is 1
[R15] has the game food? False
[R16] Path found with total cost g(x) of 24 in 0.014453649520874023s
[R13] Search nodes expanded: 58
[R13] Nodes visited: [(23, 29), (22, 29), (21, 29), (20, 29), (19, 29), (18, 29), (17, 29), (16, 29), (15, 29), (14, 29), (23, 28), (13, 29), (12, 29), (15, 28), (11, 29), (12, 28), (23, 27), (10, 29), (15, 27), (24, 29), (12, 27), (9, 29), (23, 26), (22, 26), (21, 26), (15, 26), (12, 26), (8, 29), (11, 26), (10, 26), (9, 26), (21, 25), (7, 29), (9, 25), (25, 29), (21, 24), (6, 29), (9, 24), (16, 26), (21, 23), (20, 23), (19, 23), (18, 23), (17, 23), (16, 23), (15, 23), (14, 23), (13, 23), (12, 23), (5, 29), (11, 23), (10, 23), (9, 23), (8, 23), (7, 23), (26, 29), (18, 

#### Greedy Best-First Search

In [85]:
#[R2.2] Show gbfs (greedy best first search) and Manhattan heuristic method.
!python pacman.py -l easyLayout -p Search2Agent -a fn=gbfs,heuristic=manhattanHeuristic -z .6
#!python pacman.py -l normalLayout -p Search2Agent -a fn=gbfs,heuristic=manhattanHeuristic -z .6
#!python pacman.py -l hardLayout -p Search2Agent -a fn=gbfs,heuristic=manhattanHeuristic -z .6

[Search2Agent] using function gbfs and [R16] heuristic manhattanHeuristic
[Search2Agent] using problem type PositionSearchProblem2
[R12] Initial position of pacman is (23, 29)
[R10] Final goal position is (1, 3)
[R11] Ghost Positions is/are [(9, 20), (15, 10), (26, 7)]
Number of foods is 1
[R15] has the game food? False
[R16] Path found with total cost g(x) of 24 in 0.007546901702880859s
[R13] Search nodes expanded: 26
[R13] Nodes visited: [(23, 29), (23, 28), (23, 27), (23, 26), (22, 26), (21, 26), (21, 25), (21, 24), (21, 23), (20, 23), (19, 23), (18, 23), (18, 24), (17, 23), (16, 23), (15, 23), (14, 23), (13, 23), (12, 23), (11, 23), (10, 23), (9, 23), (9, 24), (8, 23), (7, 23), (6, 23), (6, 24)]
[R13] Solution states: 25 - [(23, 29), (23, 28), (23, 27), (23, 26), (22, 26), (21, 26), (21, 25), (21, 24), (21, 23), (20, 23), (19, 23), (18, 23), (17, 23), (16, 23), (15, 23), (14, 23), (13, 23), (12, 23), (11, 23), (10, 23), (9, 23), (8, 23), (7, 23), (6, 23), (6, 24)]
[R14] Solution ac

In [86]:
##[R2.2] Show gbfs (greedy best first search) and Euclidean heuristic method.
!python pacman.py -l easyLayout -p Search2Agent -a fn=gbfs,heuristic=euclideanHeuristic -z .6
#!python pacman.py -l normalLayout -p Search2Agent -a fn=gbfs,heuristic=euclideanHeuristic -z .6
#!python pacman.py -l hardLayout -p Search2Agent -a fn=gbfs,heuristic=euclideanHeuristic -z .6

[Search2Agent] using function gbfs and [R16] heuristic euclideanHeuristic
[Search2Agent] using problem type PositionSearchProblem2
[R12] Initial position of pacman is (23, 29)
[R10] Final goal position is (1, 3)
[R11] Ghost Positions is/are [(9, 20), (15, 10), (26, 7)]
Number of foods is 1
[R15] has the game food? False
[R16] Path found with total cost g(x) of 26 in 0.008487224578857422s
[R13] Search nodes expanded: 26
[R13] Nodes visited: [(23, 29), (22, 29), (21, 29), (20, 29), (19, 29), (18, 29), (17, 29), (16, 29), (15, 29), (14, 29), (13, 29), (12, 29), (11, 29), (10, 29), (9, 29), (8, 29), (7, 29), (6, 29), (5, 29), (4, 29), (4, 28), (4, 27), (4, 26), (5, 26), (6, 26), (6, 25), (6, 24)]
[R13] Solution states: 27 - [(23, 29), (22, 29), (21, 29), (20, 29), (19, 29), (18, 29), (17, 29), (16, 29), (15, 29), (14, 29), (13, 29), (12, 29), (11, 29), (10, 29), (9, 29), (8, 29), (7, 29), (6, 29), (5, 29), (4, 29), (4, 28), (4, 27), (4, 26), (5, 26), (6, 26), (6, 25), (6, 24)]
[R14] Soluti

### Busca local (*Hill Climbing*)

In [87]:
# [R3] Show Hill Climbing Search (hcs) as Local Search method
!python pacman.py -l easyLayout -p Search2Agent -a fn=hcs -z .6
#!python pacman.py -l normalLayout -p Search2Agent -a fn=hcs -z .6
#!python pacman.py -l hardLayout -p Search2Agent -a fn=hcs -z .6

[Search2Agent] using function hcs and [R16] heuristic nullHeuristic
[Search2Agent] using problem type PositionSearchProblem2
[R12] Initial position of pacman is (23, 29)
[R10] Final goal position is (1, 3)
[R11] Ghost Positions is/are [(9, 20), (15, 10), (26, 7)]
Number of foods is 1
[R15] has the game food? False
[R16] Path found with total cost g(x) of 26 in 0.06351470947265625s
[R13] Search nodes expanded: 1710
[R13] Nodes visited: [(23, 29), (24, 29), (25, 29), (26, 29), (26, 28), (26, 27), (26, 26), (26, 25), (23, 28), (23, 27), (23, 26), (22, 26), (21, 26), (21, 25), (21, 24), (21, 23), (20, 23), (19, 23), (18, 23), (18, 22), (18, 21), (18, 20), (18, 19), (18, 18), (18, 17), (18, 16), (19, 16), (20, 16), (21, 16), (22, 16), (23, 16), (24, 16), (25, 16), (26, 16), (26, 17), (26, 18), (26, 19), (26, 20), (26, 21), (26, 22), (26, 23), (26, 24), (22, 29), (21, 29), (20, 29), (19, 29), (18, 29), (17, 29), (16, 29), (15, 29), (14, 29), (13, 29), (12, 29), (11, 29), (10, 29), (9, 29), (

# Resultados

### *easyLayout*

|Algoritmo|Otimalidade|Completude|No. nós expandidos|No. nós visitados|Complexidade|
|---|---|---|---|---|---|
|Depth-first Search                  |?|Não|253|171|$O(.)$|
|Breadth-first Search                |?|Sim|153|25 |$O(.)$|
|A* (Euclidean)                      |?|Sim|58 |25 |$O(.)$|
|Greedy Best-first Search (Euclidean)|?|Sim|26 |27 |$O(.)$|
|A* (Manhattan)                      |?|Sim|49 |25 |$O(.)$|
|Greedy Best-first Search (Manhattan)|?|Sim|26 |25 |$O(bm)$|
|Hill Climbing                       |?|Não|   |   |$O(.)$|

### *normalLayout*

|Algoritmo|Otimalidade|Completude|No. nós expandidos|No. nós visitados|Complexidade|
|---|---|---|---|---|---|
|Depth-first Search                  |?|Não|214|140|$O(.)$|
|Breadth-first Search                |?|Sim|203|30 |$O(.)$|
|A* (Euclidean)                      |?|Sim|109|30 |$O(.)$|
|Greedy Best-first Search (Euclidean)|?|Sim|48 |30 |$O(bm)$|
|A* (Manhattan)                      |?|Sim|85 |30 |$O(.)$|
|Greedy Best-first Search (Manhattan)|?|Sim|31 |32 |$O(bm)$|
|Hill Climbing                       |?|Não|   |   |$O(.)$|

### *hardLayout*

|Algoritmo|Otimalidade|Completude|No. nós expandidos|No. nós visitados|Complexidade|
|---|---|---|---|---|---|
|Depth-first Search                  |?|Sim|79 |73|$O(.)$|
|Breadth-first Search                |?|Sim|336|49|$O(.)$|
|A* (Euclidean)                      |?|Sim|263|49|$O(.)$|
|Greedy Best-first Search (Euclidean)|?|Sim|72 |49|$O(bm)$|
|A* (Manhattan)                      |?|Sim|202|49|$O(.)$|
|Greedy Best-first Search (Manhattan)|?|Sim|61 |53|$O(bm)$|
|Hill Climbing                       |?|Não|   |  |$O(.)$|

# Conclusão

# Bibliografia

**[1]** Pac-man - Wikipedia. URL: https://en.wikipedia.org/wiki/Pac-Man

**[2]** Project 1. URL: https://drive.google.com/file/d/187lgVekPC0kBmA2AjOm-KkV8Q_lFLeoH/view

**[3]** Search for AIMA 4th edition - Implementation of search algorithms and search problems for AIMA. URLs: https://github.com/aimacode/aima-python/blob/master/search4e.ipynb, https://github.com/aimacode/aima-python/blob/master/search.ipynb

**[4]** UC Berkeley CS188 Intro to AI -- Course Materials. URL: http://ai.berkeley.edu/search.html

**[5]** *tkint* documentation. URL: https://docs.python.org/3/library/tkinter.html

**[6]** Pac-Man Maze Generation. URL: https://shaunlebron.github.io/pacman-mazegen